### ISSUE TITLE
Html Parser Strips CRLF and replaces with LF

### ISSUE THREAD

**Issue by blumenshinem** [AngleSharp/AngleSharp#719]
This seems to be by design, but I'm wondering as to the intention here. I reviewed #149 and see why this should apply to XML, but perhaps it shouldn't to HTML.

I was working on a project that uses Premailer.net (which in turn uses AngleSharp) to inline some CSS for emails. Every so often, we would receive errors regarding Maximum Line Length (see RFC 2822 2.1.1 https://www.ietf.org/rfc/rfc2822.txt) for some of these emails (a few email servers were rejecting the emails we were sending out) and I ended up investigating this in depth.

Using Premailer, we'd generate the HTML as a string and use a different third party library to then send these out to interested parties. As inputs to Premailer, we would send out emails with CRLF
Example: `<html><head>etc</head>\r\n<body>yaddayadda</body></html>\r\n`. After being processed by Premailer, all newlines would be replaced with \n.

Yet this presents a problem as technically CRLF is end-of-line marker per RFC 2616 (https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html). It seems like most are lax in following this rule, where others follow it more strictly.

After investigating, we found the cause to be a result of calling NormalizeForward in BaseTokenizer in AngleSharp, which normalizes all forms of newline to LF.

While I'm not 100% confident in my analysis, I figured reaching out wouldn't hurt. It seems like one of the email clients we are using will replace LF with CRLF anyway, but for another, we have had sporadic delivery issues.

**Comment by FlorianRappl** [comment]
OK I'm sorry to hear that but the official (HTML) spec says we must use LF and frankly we need to have a uniform solution in here (otherwise it will not only eat performance but also cause confusion and bugs), so going for LF to represent line endings seems to be the way to go.

Usually, I would now advise to use a different serialization strategy / output CRLF instead of LF (this is totally in your hand, we only use LF internally and in the text nodes, which again, can be changed on serialization to, e.g., CRLF), however, since the lib that is using AngleSharp seems to be hidden from you that does not seem to be viable path.

Maybe a PR / issue in the Prmailer.net repository would help here? I think they should address this issue in serialization.

Long story short - what is the expectation of raising this issue here? Even if we would add an option for the serialization or do something else it is still Premailer.net that runs AngleSharp. So they have to upgrade + potentially expose the option.

HTH!

**Comment by blumenshinem** [comment]
Sorry, I should have done a bit more homework before commenting. You've fulfilled my expectations perfectly, which was to gain a little insight into how this works, but I should have saved you some time and investigated this myself :) (Also apologies if there is a cultural expectation to have done so, this is in fact the first open-source project I've looked into since starting my career in software)

Your answer was very helpful however, and it led me to better understand how this works internally. Implementations of IHtmlDocument should indeed use whatever internal newlines they require, in this case LF.

Premailer calls ToHtml on an IHtmlDocument and uses the extension methods for automatically resolving an appropriate formatter (hence your comments on serialization). I see now that Premailer can just implement this interface and write their own formatter, which may be the best solution. Please correct me if I'm wrong, but I'm not convinced that the HtmlMarkupFormatter shouldn't replace Symbols.LineFeed with Symbols.NewLines[0] (CRLF), something like this:

```
String IMarkupFormatter.Text(String text)
{
    var temp = Pool.NewStringBuilder();

    for (var i = 0; i < text.Length; i++)
    {
        switch (text[i])
        {
            case Symbols.Ampersand: temp.Append("&amp;"); break;
            case Symbols.NoBreakSpace: temp.Append("&nbsp;"); break;
            case Symbols.GreaterThan: temp.Append("&gt;"); break;
            case Symbols.LessThan: temp.Append("&lt;"); break;
            case Symbols.LineFeed: temp.Append(Symbols.NewLines[0]); break;
            default: temp.Append(text[i]); break;
        }
    }

    return temp.ToPool();
}
```

I'm not sure what you mean by the HTML spec saying we must use LF. HTML is a markup language separate from how text is communicated over HTTP/SMTP etc. The only concept of a new line that is pure to HTML is a line break element `<br>`. When sending HTML over SMTP or HTTP, lines must be broken by CRLF ( see RFCs or accepted answer of https://stackoverflow.com/questions/6167050/smtp-r-n-r-without-second-n-as-double-newline ). If a web server or email client wants to transmit HTML, it should abide whichever protocol it is using even if most recipients (browsers etc) will accommodate equivalents (LF or CR instead of CRLF).

Does this perhaps make more sense? If I'm way off the mark here, I don't mind closing this out or getting in touch with Premailer about perhaps writing their own formatter.

Thanks!

**Comment by blumenshinem** [comment]
Maybe I am misunderstanding something fundamental however. AngleSharp is returning HTML, as it should. Protocols like HTTP/SMTP are perhaps not concerns of this service. Perhaps then writing a new formatter would be best.

Again, thanks for your consideration

**Comment by FlorianRappl** [comment]
> Also apologies if there is a cultural expectation to have done so, this is in fact the first open-source project I've looked into since starting my career in software

No need to apologize - its all cool and good. Raising the question and discussing various solution paths is part of deriving a better understanding together. For instance, I did not know that Premailer.net is using AngleSharp - thanks for pointing this out. Wouldn't have known without your question!

> I'm not sure what you mean by the HTML spec saying we must use LF. HTML is a markup language separate from how text is communicated over HTTP/SMTP etc.

I mean that this is specified in the tokenizer model (this has nothing to do with CRLF vs LF used in HTTP or in the original source code). The HTML spec defines all possible aspects of the parsing to unify the parsing model (which is one of the greatest achievements of HTML5). So again the LF is mostly an internal representation (following the spec) and this is the reason why I distinguish between this internal representation and the serialization,, which can be controlled, e.g., by the way you described.

You can just roll your own formatter, which essentially uses the original one except for the text formatting. Here an implementation as the one described by you would do the job.

Just tell me if you need anything!

**Comment by blumenshinem** [comment]
Awesome, thank you! Again, your comments were really illuminating. Wasn't aware of this part of the html spec (I wasn't even really aware of the need for this, tbh, but can clearly see its benefit when giving it some consideration!)

As Premailer is using AngleSharp to generate HTML for emails, I think it'd make sense if they had their own formatting. I might get in touch with them or even write a formatter and see what they think. Vielen dank!

**Comment by blumenshinem** [comment]
Hey just a follow up in case you're interested.

I've decided to inform the team working on this project that they should remove newlines before using Premailer, then to meet the 998 character limit, insert them after the styles have been inlined.

I think this is the only solution. End-of-line markers for files or HTTP/SMTP really have nothing to do with HTML (as you noted above). Premailer expects HTML, nothing more, as does AngleSharp.

I arrived at this conclusion while writing the formatter and testing its behavior. Even if I were to rewrite \n as \r\n, the following line
`<html><head></head><body></body></html>\r\n`
still would parse to this anyway
`<html><head></head><body>\n</body></html>`
and thus be formatted as
`<html><head></head><body>\r\n</body></html>`

And this really makes sense when you consider protocol concerns aren't really on Premailer to accommodate anyway, even if it is practically assumed that all consumers will be tied to SMTP.

Thanks again for your help!

**Comment by FlorianRappl** [comment]
Cool thanks for following up on this! Never knew about the 998 char limit - thanks!
